califa/calibration/R3BCalifaCrystalCal2Cluster.cxx

```diff
@@ -138,193 +242,317 @@ InitStatus R3BCalifaCrystalCal2Cluster::ReInit()
+                gammaCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryId > fTotalCrystals && cryEnergy >= fProtonClusterThreshold)
+                protonCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryId <= fTotalCrystals && cryEnergy > fSaturation)
+                saturatedCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+        }
+
+        /* -------- Simulation data ------- */
+        else
+        {
+
+            if (cryEnergy >= fGammaClusterThreshold && cryEnergy < fProtonClusterThreshold)
+                gammaCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryEnergy >= fProtonClusterThreshold)
+                protonCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+        }
+    }
+
+    /* ----- Remove duplicate entries in gamma clusters ----- */
+    for (int k = 0; k < gammaCandidatesVec.size(); k++)
+    {
+        for (int s = 0; s < gammaCandidatesVec.size(); s++)
+        {
+
+            if (gammaCandidatesVec.at(k)->GetCrystalId() == (gammaCandidatesVec.at(s)->GetCrystalId() - fTotalCrystals))
+                gammaCandidatesVec.erase(gammaCandidatesVec.begin() + s);
+        }
+    }
+
+    // Sort all vectors by energy
+    std::sort(gammaCandidatesVec.begin(), gammaCandidatesVec.end(), compareByEnergy);
+    std::sort(protonCandidatesVec.begin(), protonCandidatesVec.end(), compareByEnergy);
+
+    TVector3 mother_angles, angles;
+    Double_t fRandTheta, fRandPhi;
+
+    while (protonCandidatesVec.size())
+    {
+
+        Int_t motherId = protonCandidatesVec.at(0)->GetCrystalId();
+
+        califa_candidate cluster = { motherId, vector<Int_t>(), 0.0, 0.0, 0.0, 0.0, 0.0 };
+
+        if (motherId <= fTotalCrystals)
+            mother_angles = fCalifaGeo->GetAngles(motherId) - fCalifatoTargetPos;
+
+        else
+            mother_angles = fCalifaGeo->GetAngles(motherId - fTotalCrystals) - fCalifatoTargetPos;
+
+        if (fRand)
+        {
+
+            if (motherId > fTotalCrystals)
+                fAngularDistributions[protonCandidatesVec.at(0)->GetCrystalId() - 1 - fTotalCrystals]->GetRandom2(
+                    fRandPhi, fRandTheta);
+
+            else
+                fAngularDistributions[protonCandidatesVec.at(0)->GetCrystalId() - 1]->GetRandom2(fRandPhi, fRandTheta);
+
+            cluster.theta = TMath::DegToRad() * fRandTheta;
+            cluster.phi = TMath::DegToRad() * fRandPhi;
+        }
+
+        else
+        {
+
+            cluster.theta = mother_angles.Theta();
+            cluster.phi = mother_angles.Phi();
+        }
+
+        cluster.time = protonCandidatesVec.at(0)->GetTime();
+
+        addCrystal2Cluster(&cluster, protonCandidatesVec.at(0), "proton", &usedCrystals, fTotalCrystals);
```

Comment:
As far as I understood from your code, you sort with energy for all crystalcaldata regardless they are proton- or gamma-range channels. In this scope, you are clustering hits with any crystal greater than the proton threshold. This doesn't mean the hit is "proton" range as treated in the addCrystal2Cluster function. It will result in the proton range hit not being removed properly? Or actually you can add the same algorithm for proton range as you did for gamma range in lines 298-307.

Reply:
No, he only adds low gain channels:
```
            if (cryId > fTotalCrystals && cryEnergy >= fProtonClusterThreshold)
                protonCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
```
(Personally, I keep complaining that we should replace ``cryId > fTotalCrystals`` by ``!calData.IsHighGain()`` or something.)

Reply:
> (Personally, I keep complaining that we should replace cryId > fTotalCrystals by !calData.IsHighGain() or something.)

We could put that in a parameter container for example.